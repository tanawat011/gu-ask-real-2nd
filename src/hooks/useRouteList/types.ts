import type { DeepRequired } from 'ts-essentials'
import type { WithRequired } from 'types'

export type RouteOption<Children = undefined> = {
  path?: string // for render <Route path={path} />
  fullPath?: string // for call to navigate
  element?: JSX.Element
  icon?: JSX.Element
  isIndex?: boolean
  _?: Children
}

export type RouteOptionReal<T = undefined> = WithRequired<
  RouteOption<T>,
  'path' | 'fullPath' | '_'
>

type RouteOptionChildren<T = undefined> = WithRequired<RouteOption<T>, '_'>

export type RouteMain = {
  home: RouteOptionChildren<{
    dashboard: RouteOption
  }>
  app: RouteOptionChildren<{
    todo: RouteOptionChildren<{
      dashboard: RouteOption
    }>
    article: RouteOption
  }>
  game: RouteOptionChildren<{
    sudoku: RouteOption
    ticTacToe: RouteOption
  }>
  uiComponent: RouteOptionChildren<{
    input: RouteOptionChildren<{
      button: RouteOption
      buttonGroup: RouteOption
      calendar: RouteOption
      checkbox: RouteOption
      dateAndTime: RouteOption
      radioButton: RouteOption
      rating: RouteOption
      select: RouteOption
      slider: RouteOption
      switch: RouteOption
      textField: RouteOption
      upload: RouteOption
    }>
  }>
}

type RouteAuth = {
  login: RouteOption
  register: RouteOption
  forgotPassword: RouteOption
  resetPassword: RouteOption
}

type RouteError = {
  404: RouteOption
  401: RouteOption
  403: RouteOption
  500: RouteOption
  502: RouteOption
  503: RouteOption
  504: RouteOption
}

export type AllRoutes = {
  main: RouteOptionChildren<RouteMain>
  auth: RouteOptionChildren<RouteAuth>
  error: RouteOptionChildren<RouteError>
}

export type AllRoutesReal = {
  main: RouteOptionReal<DeepRequired<RouteMain>>
  auth: RouteOptionReal<DeepRequired<RouteAuth>>
  error: RouteOptionReal<DeepRequired<RouteError>>
}
