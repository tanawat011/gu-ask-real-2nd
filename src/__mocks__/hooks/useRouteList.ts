export const authRoutes = {
  fullPath: '/auth',
  children: {
    login: {
      fullPath: '/auth/login',
    },
  },
}

export const componentRoutes = {
  children: {
    card: {
      fullPath: '/',
    },
    button: {
      fullPath: '/',
    },
    input: {
      fullPath: '/',
    },
  },
}

export const mainRoutes = {
  children: {
    home: {
      fullPath: '/',
    },
    todo: {
      fullPath: '/',
    },
    article: {
      fullPath: '/',
    },
    component: {
      fullPath: '/',
      ...componentRoutes,
    },
  },
}

export const errorRoutes = {
  children: {
    e401: {
      fullPath: '/',
    },
    e403: {
      fullPath: '/',
    },
    e404: {
      fullPath: '/',
    },
    e500: {
      fullPath: '/',
    },
    e502: {
      fullPath: '/',
    },
    e503: {
      fullPath: '/',
    },
    e504: {
      fullPath: '/',
    },
  },
}

export const allRoutes = {
  auth: authRoutes,
  mainRoutes: mainRoutes,
  error: errorRoutes,
}

export const useRouteList = () => ({
  allRoutes,
  authRoutes,
  mainRoutes,
  errorRoutes,
})
