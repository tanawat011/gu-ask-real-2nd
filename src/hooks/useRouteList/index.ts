export * from './types'
export * from './useRouteList'
