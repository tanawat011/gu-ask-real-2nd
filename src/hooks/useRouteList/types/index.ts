import type { AuthRoutesChildren } from './auth'
import type { ErrorRoutesChildren } from './error'
import type { MainRoutesChildren } from './main'
import type { WithRequired } from 'types/utilities'
import type { RenderRoute } from 'utils/route'

export type WithChildrenRoute<T> = WithRequired<RenderRoute<T>, 'children'>
export type WithJsx<T, K = JSX.Element> = {
  [key in keyof T]: K
}

// * ALL BASE ROUTES
export type AssembleRoutes = {
  auth: AuthRoutesChildren
  main: MainRoutesChildren
  error: ErrorRoutesChildren
}

export * from './auth'
export * from './error'
export * from './main'
