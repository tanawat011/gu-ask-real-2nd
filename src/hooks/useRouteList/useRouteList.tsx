import type { ObjectKeyOf } from 'types'
import type {
  AppRoutes,
  AssembleRoutes,
  AuthRoutes,
  AuthRoutesChildren,
  ErrorRoutes,
  ErrorRoutesChildren,
  InputRoutes,
  MainRoutes,
  MainRoutesChildren,
  TodoRoutes,
  UiComponentRoutes,
} from 'types/routes'

import { AuthenticationLayout } from 'layouts/Authentication'
import { ErrorLayout } from 'layouts/Error'
import { MainLayout } from 'layouts/Main'
import { ButtonUi } from 'pages/UiComponents'
import { routeGenerator } from 'utils/route'

const inputRoutes: ObjectKeyOf<InputRoutes, JSX.Element> = {
  button: <ButtonUi />,
  buttonGroup: <div>button group</div>,
  calendar: <div>calendar</div>,
  checkbox: <div>checkbox</div>,
  dateAndTime: <div>date & time</div>,
  radioButton: <div>radio button</div>,
  rating: <div>rating</div>,
  select: <div>select</div>,
  slider: <div>slider</div>,
  switch: <div>switch</div>,
  textField: <div>text field</div>,
  upload: <div>upload</div>,
}

const uiComponentRoutes: ObjectKeyOf<UiComponentRoutes, InputRoutes> = {
  input: routeGenerator<InputRoutes>('ui-component/input', inputRoutes),
}

const todoRoutes: ObjectKeyOf<TodoRoutes, JSX.Element> = {
  dashboard: <div>dashboard</div>,
}

const appRoutes: ObjectKeyOf<AppRoutes, JSX.Element | TodoRoutes> = {
  todo: routeGenerator<TodoRoutes>('app/todo', todoRoutes),
  article: <div>article</div>,
}

const mainRoutes: ObjectKeyOf<
  MainRoutes,
  JSX.Element | AppRoutes | UiComponentRoutes
> = {
  home: <div>home</div>,
  app: routeGenerator<AppRoutes, TodoRoutes>('app', appRoutes),
  uiComponent: routeGenerator<UiComponentRoutes, InputRoutes>(
    'ui-component',
    uiComponentRoutes,
  ),
}

const authRoutes: ObjectKeyOf<AuthRoutes, JSX.Element> = {
  login: <div>login</div>,
}

const errorRoutes: ObjectKeyOf<ErrorRoutes, JSX.Element> = {
  404: <div>error 404</div>,
  401: <div>error 401</div>,
  403: <div>error 403</div>,
  500: <div>error 500</div>,
  502: <div>error 502</div>,
  503: <div>error 503</div>,
  504: <div>error 504</div>,
}

export const mainRootRoute: MainRoutesChildren = {
  path: '/',
  fullPath: '/',
  element: <MainLayout />,
  children: routeGenerator<MainRoutes, UiComponentRoutes | AppRoutes>(
    '',
    mainRoutes,
  ),
}

export const authRootRoute: AuthRoutesChildren = {
  path: 'auth',
  fullPath: '/auth',
  element: <AuthenticationLayout />,
  children: routeGenerator<AuthRoutes>('auth', authRoutes),
}

export const errorRootRoute: ErrorRoutesChildren = {
  path: '',
  fullPath: '',
  element: <ErrorLayout />,
  children: routeGenerator<ErrorRoutes>('', errorRoutes),
}

type RouteList = {
  allRoutes: AssembleRoutes
  authRootRoute: AuthRoutesChildren
  mainRootRoute: MainRoutesChildren
  errorRootRoute: ErrorRoutesChildren
}

export const useRouteList = (): RouteList => {
  const allRoutes: AssembleRoutes = {
    auth: authRootRoute,
    main: mainRootRoute,
    error: errorRootRoute,
  }

  return { allRoutes, mainRootRoute, authRootRoute, errorRootRoute }
}
