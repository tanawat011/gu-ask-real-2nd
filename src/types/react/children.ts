export type Children = { children: JSX.Element }
