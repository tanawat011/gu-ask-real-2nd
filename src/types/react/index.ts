export * from './children'
export * from './reactRouterDom'
