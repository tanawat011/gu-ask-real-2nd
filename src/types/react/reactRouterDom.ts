import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
