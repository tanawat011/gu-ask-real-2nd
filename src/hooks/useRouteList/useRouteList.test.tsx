import { render } from '@testing-library/react'

import { useRouteList } from './useRouteList'

jest.mock('./useRouteList', () => ({
  ...jest.requireActual('./useRouteList'),
}))
jest.mock('layouts/Authentication')
jest.mock('layouts/Error')
jest.mock('layouts/Main')

const ExampleComponent = () => {
  const { allRoutes, authRoutes, errorRoutes, mainRoutes } = useRouteList()

  return (
    <div>
      <p>{JSON.stringify(allRoutes)}</p>
      <p>{JSON.stringify(authRoutes)}</p>
      <p>{JSON.stringify(errorRoutes)}</p>
      <p>{JSON.stringify(mainRoutes)}</p>
    </div>
  )
}

describe('useRouteList', () => {
  test('should return all routes', () => {
    expect.hasAssertions()

    const { asFragment } = render(<ExampleComponent />)

    expect(asFragment()).toMatchSnapshot()
  })
})
