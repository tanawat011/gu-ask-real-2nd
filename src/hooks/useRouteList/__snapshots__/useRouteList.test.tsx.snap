// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRouteList should return all routes 1`] = `
<DocumentFragment>
  <div>
    <p>
      {"auth":{"path":"auth","fullPath":"/auth","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"login":{"path":"login","fullPath":"/auth/login","element":{"type":"div","key":null,"ref":null,"props":{"children":"login"},"_owner":null,"_store":{}},"isIndex":true}}},"main":{"path":"/","fullPath":"/","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"home":{"path":"","fullPath":"/","element":{"type":"div","key":null,"ref":null,"props":{"children":"home"},"_owner":null,"_store":{}},"isIndex":true},"todo":{"path":"todo","fullPath":"/todo","element":{"type":"div","key":null,"ref":null,"props":{"children":"todo"},"_owner":null,"_store":{}}},"article":{"path":"article","fullPath":"/article","element":{"type":"div","key":null,"ref":null,"props":{"children":"article"},"_owner":null,"_store":{}}},"component":{"path":"component","fullPath":"/component","children":{"card":{"path":"card","fullPath":"/component/card","element":{"type":"div","key":null,"ref":null,"props":{"children":"card"},"_owner":null,"_store":{}}},"button":{"path":"button","fullPath":"/component/button","element":{"type":"div","key":null,"ref":null,"props":{"children":"button"},"_owner":null,"_store":{}}},"input":{"path":"input","fullPath":"/component/input","element":{"type":"div","key":null,"ref":null,"props":{"children":"input"},"_owner":null,"_store":{}}}}}}},"error":{"path":"","fullPath":"","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"e401":{"path":"401","fullPath":"/401","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 401"},"_owner":null,"_store":{}}},"e403":{"path":"403","fullPath":"/403","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 403"},"_owner":null,"_store":{}}},"e404":{"path":"404","fullPath":"/404","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 404"},"_owner":null,"_store":{}}},"e500":{"path":"500","fullPath":"/500","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 500"},"_owner":null,"_store":{}}},"e502":{"path":"502","fullPath":"/502","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 502"},"_owner":null,"_store":{}}},"e503":{"path":"503","fullPath":"/503","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 503"},"_owner":null,"_store":{}}},"e504":{"path":"504","fullPath":"/504","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 504"},"_owner":null,"_store":{}}}}}}
    </p>
    <p>
      {"path":"auth","fullPath":"/auth","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"login":{"path":"login","fullPath":"/auth/login","element":{"type":"div","key":null,"ref":null,"props":{"children":"login"},"_owner":null,"_store":{}},"isIndex":true}}}
    </p>
    <p>
      {"path":"","fullPath":"","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"e401":{"path":"401","fullPath":"/401","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 401"},"_owner":null,"_store":{}}},"e403":{"path":"403","fullPath":"/403","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 403"},"_owner":null,"_store":{}}},"e404":{"path":"404","fullPath":"/404","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 404"},"_owner":null,"_store":{}}},"e500":{"path":"500","fullPath":"/500","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 500"},"_owner":null,"_store":{}}},"e502":{"path":"502","fullPath":"/502","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 502"},"_owner":null,"_store":{}}},"e503":{"path":"503","fullPath":"/503","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 503"},"_owner":null,"_store":{}}},"e504":{"path":"504","fullPath":"/504","element":{"type":"div","key":null,"ref":null,"props":{"children":"error 504"},"_owner":null,"_store":{}}}}}
    </p>
    <p>
      {"path":"/","fullPath":"/","element":{"key":null,"ref":null,"props":{},"_owner":null,"_store":{}},"children":{"home":{"path":"","fullPath":"/","element":{"type":"div","key":null,"ref":null,"props":{"children":"home"},"_owner":null,"_store":{}},"isIndex":true},"todo":{"path":"todo","fullPath":"/todo","element":{"type":"div","key":null,"ref":null,"props":{"children":"todo"},"_owner":null,"_store":{}}},"article":{"path":"article","fullPath":"/article","element":{"type":"div","key":null,"ref":null,"props":{"children":"article"},"_owner":null,"_store":{}}},"component":{"path":"component","fullPath":"/component","children":{"card":{"path":"card","fullPath":"/component/card","element":{"type":"div","key":null,"ref":null,"props":{"children":"card"},"_owner":null,"_store":{}}},"button":{"path":"button","fullPath":"/component/button","element":{"type":"div","key":null,"ref":null,"props":{"children":"button"},"_owner":null,"_store":{}}},"input":{"path":"input","fullPath":"/component/input","element":{"type":"div","key":null,"ref":null,"props":{"children":"input"},"_owner":null,"_store":{}}}}}}}
    </p>
  </div>
</DocumentFragment>
`;
